agent/lib/kontena/service_pods/lifecycle_hook_manager.rb

```diff
@@ -0,0 +1,159 @@
+        return if pre_start_hooks.size == 0
+
+        pre_start_hooks.each do |hook|
+          begin
+            info "running pre_start hook: #{hook['cmd']}"
+            service_config = config_container(service_pod.dup, hook['cmd'])
+            service_config['name'] = "#{service_config['name']}-#{SecureRandom.urlsafe_base64(5)}"
+            service_container = create_container(service_config)
+            log_hook_output(service_container.id, "running pre_start hook: #{hook['cmd']}", 'stdout')
+            service_container.tap(&:start).attach
+            if service_container.state['ExitCode'] != 0
+              raise "Failed to execute pre_start hook: #{hook['cmd']}, exit code: #{service_container.state['ExitCode']}"
+            end
+          ensure
+            cleanup_container(service_container) if service_container
+          end
+        end
+        true
+      rescue => exc
+        log_service_pod_event("service:create_instance", exc.message, Logger::ERROR)
+        raise exc
+      end
+
+      # @param [Docker::Container] service_container
+      # @return [Boolean]
+      def on_post_start(service_container)
+        hooks_for('post_start').each do |hook|
+          info "running post_start hook: #{hook['cmd']}"
+          command = build_cmd(hook['cmd'])
+          log_hook_output(service_container.id, "running post_start hook: #{hook['cmd']}", 'stdout')
+          _, _, exit_code = service_container.exec(command) { |stream, chunk|
+            log_hook_output(service_container.id, chunk, stream)
+          }
+          if exit_code != 0
+            raise "Failed to execute post_start hook: #{hook['cmd']} (exit code: #{exit_code}"
+          end
+        end
+        true
+      rescue => exc
+        log_service_pod_event("service:create_instance", exc.message, Logger::ERROR)
+        error exc.message
+        false
+      end
+
+      # @param [Docker::Container] service_container
+      # @return [Boolean]
+      def on_pre_stop(service_container)
+        hooks_for('pre_stop').each do |hook|
+          info "running pre_stop hook: #{hook['cmd']}"
+          command = build_cmd(hook['cmd'])
+          log_hook_output(service_container.id, "running pre_stop hook: #{hook['cmd']}", 'stdout')
+          _, _, exit_code = service_container.exec(command) { |stream, chunk|
+            log_hook_output(service_container.id, chunk, stream)
+          }
+          if exit_code != 0
+            raise "Failed to execute pre_stop hook: #{hook['cmd']} (exit code: #{exit_code}"
+          end
+        end
+        true
+      rescue => exc
+        log_service_pod_event("service:create_instance", exc.message, Logger::ERROR)
+        error exc.message
+        false
+      end
+
+      # @param [String] type
+      # @return [Array<Hash>]
+      def hooks_for(type)
+        hooks = service_pod.hooks.select { |h|
+          h['type'] == type && !h['oneshot'] || (h['oneshot'] && !oneshot_cache.include?(h))
```

Comment:
This is also returning `oneshot` `post_start` hooks for `type = 'pre_start'`, so they get run as `pre_start` hooks:

```yaml
stack: test/hooks-oneshot
services:
  redis:
    image: redis
    hooks:
      post_start:
        - name: test-first
          cmd: date && echo "first"
          instances: '*'
          oneshot: false
        - name: test-echo1
          cmd: date && echo "pre-sleep"
          instances: '*'
          oneshot: true
        - name: test-sleep1
          cmd: sleep 5
          instances: '*'
          oneshot: true
        - name: test-echo2
          cmd: date && echo "mid-sleep"
          instances: '*'
          oneshot: true
        - name: test-sleep2
          cmd: sleep 5
          instances: '*'
          oneshot: true
        - name: test-echo3
          cmd: date && echo "post-sleep"
          instances: '*'
          oneshot: true
        - name: test-last
          cmd: date && echo "last"
          instances: '*'
          oneshot: false
```

```
I, [2017-09-20T11:44:36.983241 #1]  INFO -- Kontena::ServicePods::Creator: creating service: hooks-oneshot.redis-1
I, [2017-09-20T11:44:36.983601 #1]  INFO -- Kontena::Workers::ImagePullWorker: pulling image: redis:latest
I, [2017-09-20T11:44:39.707346 #1]  INFO -- Kontena::Workers::ImagePullWorker: pulled image: redis:latest
I, [2017-09-20T11:44:39.883092 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running pre_start hook: date && echo "pre-sleep"
I, [2017-09-20T11:44:41.903369 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running pre_start hook: sleep 5
I, [2017-09-20T11:44:49.354710 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running pre_start hook: date && echo "mid-sleep"
I, [2017-09-20T11:44:51.535713 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running pre_start hook: sleep 5
I, [2017-09-20T11:44:58.802024 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running pre_start hook: date && echo "post-sleep"
D, [2017-09-20T11:45:01.448684 #1] DEBUG -- Kontena::ServicePods::Creator: creating container: hooks-oneshot.redis-1
D, [2017-09-20T11:45:01.543426 #1] DEBUG -- Kontena::ServicePods::Creator: container created: hooks-oneshot.redis-1
I, [2017-09-20T11:45:01.962542 #1]  INFO -- Kontena::ServicePods::Creator: service started: hooks-oneshot/redis-1
I, [2017-09-20T11:45:01.967796 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "first"
I, [2017-09-20T11:45:02.175671 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "last"
```

Reply:
Should be fixed now.